rails/react-components/src/library/components/permission-forms-v2/create-edit-permission-form.tsx

```diff
@@ -0,0 +1,67 @@
+import React, { useState } from "react";
+import { IProject, IPermissionFormFormData, CurrentSelectedProject, IPermissionForm } from "./permission-form-types";
+
+import css from "./style.scss";
+
+interface CreateEditPermissionFormProps {
+  existingFormData?: IPermissionForm;
+  currentSelectedProject?: CurrentSelectedProject;
+  projects: IProject[];
+  onFormCancel: () => void;
+  onFormSave: (newForm: IPermissionFormFormData) => void;
+}
+
+export const CreateEditPermissionForm = ({ projects, currentSelectedProject, existingFormData, onFormSave, onFormCancel }: CreateEditPermissionFormProps) => {
+  const [formData, setFormData] = useState<IPermissionFormFormData>({
+    id: existingFormData?.id || undefined,
+    name: existingFormData?.name || "",
+    project_id: existingFormData?.project_id || (currentSelectedProject ? Number(currentSelectedProject) : ""),
+    url: existingFormData?.url || ""
+  });
+
+  const handleFormInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setFormData({ ...formData, [e.target.name]: e.target.value });
```

Comment:
Hm, I didn't write this code (I guess git diff got confused because of the rename), so I didn't give it much thought before.

Yes, the callback form of the setter is always a safer and better option. Theoretically, typing in the text input might happen fast enough for updates to be batched and/or for their order not to be maintained, but actually, I'd be pretty surprised to see that happen in practice here.

Or do you mean some other issue?

I can definitely change that to the callback form anyway.

Reply:
Done in https://github.com/concord-consortium/rigse/pull/1320/commits/cef013006da0480f4cc10e7f0dfa30b71059a003